src/librustc_mir/transform/mod.rs

```diff
@@ -440,6 +444,25 @@ fn run_optimization_passes<'tcx>(
         // optimizations that run before, that might be harder to do on the state machine than MIR
         // with async primitives.
         &generator::StateTransform,
+    ];
+    #[rustfmt::skip]
+    run_passes(
```

Comment:
So the point of this is to get a different phase the separate set of optimizations?

I think this would be more readable and less repetetive if `run_pass` took something more like `&[(MirPhase, &[MirPass])]`. Then it could all be one `run_pass` call.

Reply:
That's a nontrivial change. There's funky code like `if body.phase >= mir_phase { return; }` which would mean that if I have two `MirPhase::Optimized` (like we have right now with the actual optimizations and the pre-codegen transformations), we'd skip the pre-codegen transformations. Everything would become very fragile, and I'd rather not touch this logic.

Reply:
Repeating this huge `run_passs` invocation also seems rather fragile, and in particular it makes the "pass manager" (aka the `run_optimization_passes` function) basically unreadable to me.

Of course this means `run_passes` cannot assume any more that all passes belong to the same phase, so that `return` would become a `continue` in the loop.

Reply:
Oh but also the code changed since I last saw it, `post_borrowck_cleanup` used to be together with the rest. So it's already much harder to follow than in the past...

Reply:
> Of course this means `run_passes` cannot assume any more that all passes belong to the same phase, so that `return` would become a `continue` in the loop.

That won't help either, as two `pass_group`'s with the same `MirPhase` (e.g. `optimizations` and `pre_codegen`) will cause only the first pass group to be applied. Either we force one `MirPhase` per `pass_group` or I have no idea how to solve this without a major refactoring.

Reply:
Maybe we can restructure the code to consist first of declaring all the lists of phases (with suitable comments), and then at the bottom the two `run_pass` calls. The comments should be such that one doesn't have to read the `run_pass` calls and still know what goes on.

Reply:
Done that, that's indeed much better